lncli: add command to create new macaroon

guggero opened the pull request:
As discussed in #1147 there is a need to create macaroons with custom permissions set since the three existing macaroon files `admin.macaroon`, `invoice.macaroon` and `readonly.macaroon` aren't fine-grained enough.

A new gRPC method named `NewMacaroon` is added:
* Introduces a new permission entity named `macaroon` (could be used for macaroon based RPCs mentioned in #291 too).
* `write` access to the entity macaroon is necessary to call the method NewMacaroon.
* The `admin.macaroon` gets `write` access to the entity `macaroon`.
* A new command is added to lncli named `newmacaroon` that calls this gRPC method.
* As parameters to the `NewMacaroon` method a list of entity/action pairs for the allowed operations can be passed.

Example:
```bash
lncli newmacaroon --permission=invoices/write --permission=invoices/read --save_to=~/.lnd/custom-invoice.macaroon --timeout=10
```
Creates a macaroon that is valid for reading and writing invoices during the next 10 seconds.

Closes #283, #1147, #3516.

**NOTE** for release notes: Users will need to delete or move their `admin.macaroon`, `readonly.macaroon` and `invoices.macaroon` before starting `0.9`, otherwise they won't get regenerated macaroons that have the required permission (`macaroon:generate`) to mint custom macaroons.

vegardengen commented:
This one actually conflicts with #1147 

Would it make sense to make this into one common PR? Or merge one, then rebase/fix conflicts after that?

aakselrod commented:
These changes are looking OK from a quick glance. Will start testing after #1147 is merged and this is rebased.

guggero commented:
I've rebased and added the same `--save_to` parameter as in #1147.

joostjager commented on cmd/lncli/commands.go:
```diff
@@ -4016,3 +4030,255 @@ func restoreChanBackup(ctx *cli.Context) error {
+	}
+
+	return nil
+}
```
This function could use some method extraction perhaps.

joostjager commented on cmd/lncli/commands.go:
```diff
@@ -4016,3 +4030,255 @@ func restoreChanBackup(ctx *cli.Context) error {
 	return nil
 }
+
+var newMacaroonCommand = cli.Command{
```
To prevent the `commands.go` file from growing larger and larger, we started placing new commands in their own file. See `cmd_*.go`

joostjager commented on rpcserver.go:
```diff
@@ -5233,3 +5241,37 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+
+	// Bake new macaroon with the given permissions and send it binary
+	// serialized and hex encoded to the client.
+	newMac, err := r.server.macService.Oven.NewMacaroon(
```
If you'd be following up this PR with more granular macaroon baking on the individual call level, how would that map to `bakery.Op`s?

guggero replied:
My idea for the follow-up PR would be that we would use a constant for the `entity` (for example `"rpc"`) and the full RPC URL (as seen in `rpcserver.go`, for example `"/lnrpc.Lightning/SendCoins"`) as the `action`. This would result in quite large macaroons but would offer most flexibility.

joostjager replied:
Have we already heard of a use case where our partitioning is insufficient and this fine granularity is required?

joostjager commented:
During testing I noticed that all of the main macaroons (not just admin.macaroon) need to be removed in order to trigger regeneration. Idk if there is a reason for that.

wpaulino commented on cmd/lncli/commands.go:
```diff
@@ -4016,3 +4030,255 @@ func restoreChanBackup(ctx *cli.Context) error {
+	Category: "Macaroons",
+	Usage: "Bakes a new macaroon with the provided list of " +
+		"permissions and restrictions",
+	ArgsUsage: "--permission=entity/action [--format= | --save_to=] " +
```
I would prefer something like `--permissions=onchain:write,offchain:read`, thoughts?

wpaulino commented on cmd/lncli/commands.go:
```diff
@@ -4016,3 +4030,255 @@ func restoreChanBackup(ctx *cli.Context) error {
+				"valid before it times out",
+		},
+		cli.StringFlag{
+			Name:  "ip_address",
```
Perhaps leave this and `timeout` out since they can already be specified with the `lncli` flags?

guggero replied:
The IP address and timeout are already provided by `lncli`, yes. But the user can never get access to those attenuated macaroons since they are only ever used to be sent over the wire. So I think it's still useful to be able to mint and directly restrict a macaroon with this command. Especially when we can do it client side with minimal effort.

wpaulino commented on cmd/lncli/commands.go:
```diff
@@ -4016,3 +4030,255 @@ func restoreChanBackup(ctx *cli.Context) error {
+	Or it can be saved directly to a file using the --save_to argument.
+
+	A permission is a tuple of an entity and an action.
+	The following entities are available in lnd:
```
This can quickly become outdated. Perhaps it's worth introducing a flag/separate command to list the entities and actions available? Exposing which ones all available RPCs require would be nice too, at the moment users have to read the code. Sounds like follow-up work to me though.

guggero commented:
Thank you @joostjager and @wpaulino for the reviews!
I addressed everything unless explicitly commented above.
As discussed offline, I moved the list of permissions that exist to the server side and removed it from the command documentation text. The full list of supported actions and entities is returned in the error message if a user specifies something incorrect.

Will introduce a RPC call to get all supported permissions in the follow-up PR that also introduces fine-grained permissions on individual RPC method level.

joostjager commented on cmd/lncli/cmd_new_macaroon.go:
```diff
@@ -0,0 +1,237 @@
+	// argument or through a flag.
+	switch {
+	case ctx.IsSet("permissions"):
+		permissions = strings.Split(ctx.String("permissions"), ",")
```
maybe remove the flag, because now the behavior is maybe not as expected when both the flag and the positional args are set.

joostjager commented on cmd/lncli/cmd_new_macaroon.go:
```diff
@@ -0,0 +1,237 @@
+		}
+		entity, action := tuple[0], tuple[1]
+		if entity == "" {
+			return fmt.Errorf("invalid permission. entity cannot " +
```
include the `permission` string in error?

joostjager commented on cmd/lncli/cmd_new_macaroon.go:
```diff
@@ -0,0 +1,237 @@
+		printJSONBytes(jsonBytes)
+
+	// Defaults to format 'hex'.
+	default:
```
could make it an explicit case to not rely on the comment to understand

joostjager commented on lnrpc/README.md:
```diff
@@ -120,7 +120,13 @@ description):
+       globally, or a particular channel.
+  * ForwardingHistory
+     * ForwardingHistory allows the caller to query the htlcswitch for a
+       record of all HTLCs forwarded.
```
ForwardingHistory doc could have been a separate commit

joostjager commented on rpcserver.go:
```diff
@@ -151,6 +151,10 @@ var (
 		},
+		{
+			Entity: "macaroon",
+			Action: "write",
```
`write` or `generate`?

joostjager commented on rpcserver.go:
```diff
@@ -5243,3 +5256,65 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+	rpcsLog.Debugf("[newmacaroon]")
+
+	var (
+		validActions  = []string{"read", "write", "generate"}
```
Add TODO to remind us that we need to fetch the data from somewhere eventually.

joostjager commented on rpcserver.go:
```diff
@@ -5243,3 +5256,65 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+	// Map permission struct used by gRPC to the one used by the bakery.
+	requestedPermissions := make([]bakery.Op, len(req.Permissions))
+	for idx, permission := range req.Permissions {
+		requestedPermissions[idx] = bakery.Op{
```
Could do the validation in here to only have one loop

joostjager commented on rpcserver.go:
```diff
@@ -5243,3 +5256,65 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+	)
+
+	// Validate the requested permissions first.
+	for _, op := range req.Permissions {
```
Also return the error listing the permissions if the permission list is empty?

joostjager commented on rpcserver.go:
```diff
@@ -396,6 +400,10 @@ func mainRPCServerPermissions() map[string][]bakery.Op {
 		}},
+		"/lnrpc.Lightning/NewMacaroon": {{
+			Entity: "macaroon",
+			Action: "write",
```
action generate?

joostjager commented on rpcserver.go:
```diff
@@ -5243,3 +5256,76 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+
+	rpcsLog.Debugf("[newmacaroon]")
+
+	var (
```
Could define these lists globally so that they aren't allocated every time. They can then also be close to the permission lists in this file.

joostjager commented on rpcserver.go:
```diff
@@ -5243,3 +5256,76 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+			"onchain", "offchain", "address", "message",
+			"peers", "info", "invoices", "signer", "macaroon",
+		}
+		stringInSlice = func(a string, slice []string) bool {
```
maybe just make it a normal function?

joostjager commented on rpcserver.go:
```diff
@@ -5243,3 +5256,76 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+		if op.Entity == "" || !stringInSlice(op.Entity, validEntities) {
+			return nil, fmt.Errorf("invalid permission entity. "+
+				"supported entities are %v", validEntities)
+		}
```
So we now do allow combinations that aren't used? Like peers/generate. Probably ok.

guggero replied:
Yes, we have no way of dynamically checking this at the moment. This will be addresses in the follow-up PR where the permission lists are refactored.

joostjager commented on cmd/lncli/cmd_new_macaroon.go:
```diff
@@ -0,0 +1,226 @@
+	// The permissions need to be set as the only positional argument that
+	// is mandatory.
+	if !args.Present() {
+		return fmt.Errorf("permissions not specified")
```
should we allow this to let the server return its error?

guggero replied:
We can't run into this, the first check (`ctx.NArg()`) already catches this condition.
I think we usually just print the help text if no arguments or flags are specified. So with the CLI we can't really hit that "permission list cannot be empty" error.

joostjager commented on rpcserver.go:
```diff
@@ -5243,3 +5276,56 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+	// the bakery.
+	requestedPermissions := make([]bakery.Op, len(req.Permissions))
+	for idx, op := range req.Permissions {
+		if op.Action == "" || !stringInSlice(op.Action, validActions) {
```
`==""` is handled already by `stringInSlice`?

joostjager commented on cmd/lncli/cmd_new_macaroon.go:
```diff
@@ -0,0 +1,221 @@
+	// Split will always give us at least one slice. So we'll check each
+	// entry if it's a valid entity:action tuple. The content itself is
+	// validated server side. We just make sure we can parse it correctly.
+	permissions = strings.Split(args.First(), ",")
```
is the `,` better than using separate args (space separated)?

guggero replied:
Hmm.. Yes, that is nicer and easier to handle.

joostjager commented on cmd/lncli/cmd_new_macaroon.go:
```diff
@@ -0,0 +1,221 @@
+
+	// Show command help if no arguments or flags are provided. Also show
+	// help if there are more arguments than we expect.
+	if (ctx.NArg() == 0 || ctx.NArg() > 1) && ctx.NumFlags() == 0 {
```
`ctx.NArg() != 1` ?

joostjager replied:
If there is one flag and no args, it won't show the help?

guggero replied:
Fixed.

wpaulino commented on rpcserver.go:
```diff
@@ -5243,3 +5276,56 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
+
+	// Bake new macaroon with the given permissions and send it binary
+	// serialized and hex encoded to the client.
+	newMac, err := r.macService.Oven.NewMacaroon(
```
This will panic if macaroons are disabled through `--no-macaroons`, so we should error early if this is the case (at the top of the RPC before any validation).

guggero replied:
Nice catch, thanks!

guggero commented:
@Roasbeef asked me to rename the service to `BakeMacaroon`.
